packages/parser-javascript/src/walk.ts

```diff
@@ -0,0 +1,222 @@
+        }
+
+        /**
+         * Every time a hint calls to walk.(simple|ancestor|full|fullAncestor), it is going to
+         * execute this method, storing in a WalkArray object all the NodeVistors or Callbacks
+         * the hints are defining for the walk method.
+         *
+         * This will allow later generate our custom NodeVisitor(s) or callback to call the
+         * real `acorn-walk` method, so we just need to walk once for each Key (node + base + state)
+         * and method.
+         *
+         * E.g:
+         * For the script1.js
+         *
+         * hint 1: call to walk.simple(ast, {
+         *     CallExpression(node) { // this is a NodeVisitor
+         *         // any code here
+         *     }
+         * });
+         *
+         * hint 2: call to walk.simple(ast, {
+         *     CallExpression(node) { // This is a NodeVisitor
+         *         // any other code here
+         *     }
+         * });
+         *
+         * hint 3: call to walk.simple(ast, {
+         *     Literal(node) { // This is a NodeVisitor
+         *         // any code for Literal.
+         *     }
+         * });
+         *
+         * hint 4: call to walk.full(ast, (node) => {
+         *   // Callback code here.
+         * });
+         *
+         * These hints will create two WalkArray, one for the method `simple` and another one
+         * for the method `full`.
+         *
+         * The WalkArray object for `simple` will have a map with 2 entries, the key of the first entry
+         * will be `CallExpression` and the content for that entry will be an array with 2 NodeVisitors,
+         * one from `hint 1`, and another from the `hint 2`. The key for the second entry will be `Literal`
+         * and the content for that entry will be an array with 1 NodeVisitor the one from `hint 3`
+         *
+         * The WalkArray object for `full` will hava a map with 1 entry, the key of that entry will
+         * be `callbacks`, and the content for that entry will be an array with 1 function. That function
+         * is the function defined in `hint 4`
+         */
+        return (node: Node, visitorsOrCallback: NodeVisitor | Function, base?: NodeVisitor, state?: any) => {
+            let currentVisitors = getCurrentVisitorsOrCallback(walkArrays[methodName], node, base, state);
+
+            if (!currentVisitors) {
+                currentVisitors = new Map();
+                walkArrays[methodName].push([{ base, node, state }, currentVisitors]);
+            }
+
+            if (typeof visitorsOrCallback === 'function') {
+                // `full` and `fullAncestor` only track an array of callbacks.
+                const name = defaultCallbacksProperty;
+                const visitorCallbacks = currentVisitors.get(name) || [];
+
+                visitorCallbacks.push(visitorsOrCallback);
+                currentVisitors.set(name, visitorCallbacks);
+
+                return Promise.resolve();
```

Comment:
Can't we just do `return`? The code doesn't seem to return any "real"  `Promise`  🤔
Probably worth a comment if there's a particular reason.

Reply:
Yep, this can just be a `return` (as can the similar `return` at the end of this function). This was left over from a different experiment with registering post-walk callbacks.